### PR TITLE
Fix unwind alignment

### PR DESCRIPTION
**Pull request by triplef**
Updated _Unwind_Exception structs to match latest revisions from libunwind:

- ARM version is using `__attribute__((__aligned__(8)))`:  
https://github.com/llvm/llvm-project/blob/master/libunwind/include/unwind.h#L107

- Itanium version is using `__attribute__((__aligned__))`, `uintptr_t` for private fields, and an added `reserved` field to be explicit about alignment:
https://github.com/llvm/llvm-project/blob/master/libunwind/include/unwind.h#L123-L143

This was previously submitted as https://github.com/pathscale/libcxxrt/pull/54.

### CODE REVIEW

**Comment by davidchisnall** [comment]
@emaste, this has come up in the context of C++ exception interop on Android.  It's an issue there because the alignment and size of these control structures must match between the generic unwinder, the C++ runtime library and the Objective-C runtime library.  I believe the versions in libcxxrt currently match the ones in the GNU libunwind (hence, they work on FreeBSD), but not LLVM's libunwind (or possibly not on all platforms).

I think the changes here are strictly more correct, but are a change to the ABI.  This probably matters only for libunwind / libgcc_s, libcxxrt and libobjc on FreeBSD, but we should avoid breakage, so probably want to add some `#ifdef`s around this and think about how to roll out the changes.

**Comment by emaste** [comment]
@davidchisnall these commits change only itanium and 32-bit arm AFAICT? No concerns for Itanium, and for 32-bit arm as a Tier-2 arch there's not much concern for backwards compatibility as long as the change is sufficiently documented.

**Comment by emaste** [comment]
Oh nvm about the itanium comment, pondering...

**Comment by davidchisnall** [comment]
Applying this fix in libobjc broke FreeBSD in CI, but I don't know what happens if you apply the fix in libcxxrt but not libobjc.  I don't think the weaker alignment is actually a problem (for most architectures, at least), it's just that it needs to be consistent.

The AArch32 changes are probably fine for FreeBSD, if we don't mind ABI breaking changes there (though I think NetBSD would probably be unhappy. @jsonn?).  

I'd also be interested in whether these changes break compatibility with the GCC unwinder in any subtle ways.

**Comment by jsonn** [comment]
We've been using the equivalent in our own copy for a long time for Itanium unwinding. I think the ABI impact is small to non-existing since libunwind controls the allocation. The primary difference is whether the compiler assumes to have access to SSE2 instructions and the like.

**Comment by triplef** [comment]
Sorry for the noise... I accidentally updated this branch while working on #2, but now it’s back to the original change.

**Comment by davidchisnall** [comment]
@triplef, I think I am slightly confused across the two PRs.  Does this one change the size of the C++ exception structure (thus changing the ABI)?  If so, we probably need it to be conditionally compiled.  If not, then I'm happy to commit it as-is.

**Comment by jsonn** [comment]
@davidchisnall Strictly speaking, it is an ABI change if exceptions are declared as static objects or put into other objects. But noone is supposed to create instances of the structure but ask libunwind for them. As such, the stricter alignment should not change well defined code.

**Comment by davidchisnall** [comment]
@jsonn, the Itanium C++ ABI and the ARM ABI both embed these structures in their `__cxa_exception` structure, as does the Objective-C runtime.  The `__cxa_exception` structure, returned by libcxxrt via an Itanium-ABI-defined structure has a layout that is part of the public ABI.

**Comment by triplef** [comment]
> @triplef, I think I am slightly confused across the two PRs.

Do you mean this one and the equivalent PR in libobjc2 (https://github.com/gnustep/libobjc2/pull/129)? They contain the following changes:

1. both make the identical changes to the _Unwind_Exception structure alignment on ARM and Itanium. This was done to match the structure with how it is defined in the latest libunwind.
2. the libobjc2 change also updates the __cxa_exception structure to add the referenceCount field, which is [already present](https://github.com/libcxxrt/libcxxrt/blob/master/src/cxxabi.h#L90) in libcxxrt. So with that PR the __cxa_exception structures are in sync again between libcxxrt and libobjc2.

Hope that helps.

**Comment by davidchisnall** [comment]
I am a bit unsure why applying the libobjc2 change but not this one then causes CI to fail on FreeBSD there - I would expect that to be fixing ABI breakage caused by libobjc2 having the wrong definition for this field, not introducing incompatibilities.

**Comment by triplef** [comment]
@davidchisnall is that CI using GNU’s libunwind? Then my guess is it’s because with these PRs the _Unwind_Exception structure size no longer matches GNU’s libunwind, probably because GNU libunwind does not have the `reserved` field. Is there a way to detect which libunwind is used at compile time?

**Comment by davidchisnall** [comment]
@emaste, it's using Cirrus' freebsd-12-0-release-amd64 image.  Is that still using GCC's libunwind, or LLVM's?  

Perhaps also relevant, if it's using GCC's and this is not quite the same ABI as LLVM's, then when we switch to LLVM's we'll need to either preserve the ABI or do a .so bump.

@triplef, probably the easiest way to check at compile time is to have a CMake check that tries both versions and sees which one crashes...

**Comment by emaste** [comment]
LLVM_LIBUNWIND was set by default in https://github.com/freebsd/freebsd/commit/d20793840b5b74acebe80ec710522f7386b452cf so before 12.0

**Comment by triplef** [comment]
I downloaded FreeBSD-12.0-release-amd64 and it contains unwind.h with the following definition:

```c
struct _Unwind_Exception
  {
    uint64_t exception_class;
    _Unwind_Exception_Cleanup_Fn exception_cleanup;
    unsigned long private_1;
    unsigned long private_2;
  } __attribute__((__aligned__));
```

It already has the alignment attribute, and as the CI is on 64 bit this should match the definition from this PR in size (32 bytes).

I also took a look at their `__cxa_allocate_exception()` in libcxxrt.so.1, and from what I can tell it is using sizeof(__cxa_exception) = 128 bytes, which would match the struct size with the added "referenceCount" field from the libobjc2 PR. Here’s the disassembled pseudocode from Hopper:

```c
int __cxa_allocate_exception(int arg0) {
    rax = sub_169e0(arg0 - 0xffffffffffffff80);
    rax = rax - 0xffffffffffffff80;
    return rax;
}
```

So in summary, I still have no good idea why the FreeBSD CI is failing...

**Comment by triplef** [comment]
I’ve reproduced the issue on a FreeBSD 12.0 Vagrant box and I’ll try to come up with a change that doesn’t break existing setups.

**Comment by triplef** [comment]
Ok so with libcxxrt provided by FreeBSD and libobjc2 master built from source all tests pass, as expected.

However, with libcxxrt master and libobjc2 master both built from source, the "CXXExceptions" tests fail. This is because the sizes of the two structs currently differ between the latest libobjc2 and libcxxrt master branches (because libobjc2 doesn’t have the "referenceCount" field in __cxa_exception):

```
libcxxrt __cxa_exception:   sizeof 120, alignof 8
libcxxrt _Unwind_Exception: sizeof 32,  alignof 8
libobjc2 __cxa_exception:   sizeof 112, alignof 16
libobjc2 _Unwind_Exception: sizeof 32,  alignof 16
```

Using the pull requests of both projects re-aligns the  structs and all tests pass:

```
libcxxrt __cxa_exception:   sizeof 128, alignof 16
libcxxrt _Unwind_Exception: sizeof 32,  alignof 16
libobjc2 __cxa_exception:   sizeof 128, alignof 16
libobjc2 _Unwind_Exception: sizeof 32,  alignof 16
 ```

The easiest would be if we can live without backwards-compatibility with the old libcxxrt shipping with FreeBSD 12.0 (e.g. with an .so bump). In that case I think the two PRs can be merged as-is as they fix a current misalignment between libcxxrt and libobjc2 (and the libobjc2 CI should be updated to use the latest libcxxrt to make it pass).

If backwards-compatibility with the old version of libcxxrt is wanted we’d have to somehow poke the libcxxrt.so being used to figure out what size of __cxa_exception it is using, which would be one of the following:

* 112 bytes (libcxxrt from FreeBSD 12.0)
* 120 bytes (current libcxxrt master)
* 128 bytes (libcxxrt with this PR)

However I am not sure how exactly that could be accomplished. We can call `__cxa_allocate_exception(0)` to get a memory block with one of the sizes above, but is there any reliable way to find out its size? Writing to it until the app crashes would probably not be reliable enough, unless there’s something like "guard malloc" we can use.

**Comment by davidchisnall** [comment]
libobjc2 has now been fixed to work with the version of libcxxrt in this PR and with the version in FreeBSD/STABLE12.  I hope FreeBSD didn't import in the middle, as that version had weaker alignment for the exception class than is required...
